alns/select/MABSelector.py

```diff
@@ -0,0 +1,201 @@
+
+    References
+    ----------
+    .. [1] Emily Strong, Bernard Kleynhans, & Serdar Kadioglu (2021).
+           MABWiser: Parallelizable Contextual Multi-armed Bandits.
+           Int. J. Artif. Intell. Tools, 30(4), 2150021:1–2150021:19.
+    """
+
+    def __init__(
+        self,
+        scores: List[float],
+        num_destroy: int,
+        num_repair: int,
+        learning_policy: "LearningPolicy",
+        neighborhood_policy: Optional["NeighborhoodPolicy"] = None,
+        seed: Optional[int] = None,
+        op_coupling: Optional[np.ndarray] = None,
+        **kwargs,
+    ):
+        if not MABWISER_AVAILABLE:
+            raise ImportError("MABSelector requires the MABWiser library. ")
+
+        super().__init__(num_destroy, num_repair, op_coupling)
+
+        if any(score < 0 for score in scores):
+            raise ValueError("Negative scores are not understood.")
+
+        if len(scores) < 4:
+            # More than four is OK because we only use the first four.
+            raise ValueError(f"Expected four scores, found {len(scores)}")
+
+        # forward the seed argument if not null
+        if seed is not None:
+            kwargs["seed"] = seed
+
+        # the set of valid operator pairs (arms) is equal to the cartesian
+        # product of destroy and repair operators, except we leave out any
+        # pairs disallowed by op_coupling
+        arms = [
+            f"{d_idx}_{r_idx}"
+            for d_idx, r_idx in itertools.product(
+                range(num_destroy), range(num_repair)
+            )
+            if op_coupling is None or op_coupling[d_idx, r_idx]
+        ]
+        self._mab = MAB(
+            arms,
+            learning_policy,
+            neighborhood_policy,
+            **kwargs,
+        )
+        self._scores = scores
+
+        def extract_context(state: ContextualState):
+            if self._mab.is_contextual is False:
+                return None
+            else:
+                try:
+                    return np.atleast_2d(state.get_context())
+                except AttributeError:
+                    raise ValueError(
+                        "Contextual MAB requires a `get_context` function on"
+                        " the state object"
+                    )
+
+        self._context_extractor = extract_context
+
+    @property
+    def scores(self) -> List[float]:
+        return self._scores
+
+    @property
+    def mab(self) -> "MAB":
+        return self._mab
+
+    def __call__(  # type: ignore[override]
+        self,
+        rnd_state: RandomState,
+        best: ContextualState,
+        curr: ContextualState,
+    ) -> Tuple[int, int]:
+        """
+        Returns the (destroy, repair) operator pair from the underlying MAB
+        strategy
+        """
+        if self._mab._is_initial_fit:
+            prediction = self._mab.predict(
+                contexts=self._context_extractor(curr)
+            )
+            return arm2ops(prediction)
```

Comment:
```suggestion
            has_context = self._mab.is_contextual
            context = np.atleast_2d(curr.get_context()) if has_context else None
            prediction = self._mab.predict(contexts=context)
            return arm2ops(prediction)
```
And similarly below. Do we need `self._context_extractor`?